### PR TITLE
Add Deep SAD (Deep Semi-Supervised Anomaly Detection) detector

### LINKED ISSUE #709: Add Deep SAD (semi-supervised deep anomaly detection, ICLR 2020)
### Summary

PyOD has `DeepSVDD` (the unsupervised one-class deep detector) but not its semi-supervised counterpart, **Deep SAD** (Ruff et al., *Deep Semi-Supervised Anomaly Detection*, ICLR 2020). I'd like to contribute it — opening this first to check you're open to it before writing any code.

### Why Deep SAD fits PyOD

- It's the natural label-using sibling of the `DeepSVDD` already in PyOD: the same one-class objective, but it also exploits the (usually few) labeled anomalies/normals that are common in practice — exactly the regime a lot of real anomaly-detection use cases sit in (a large unlabeled stream plus a handful of confirmed labels).
- It's canonical and heavily cited (ICLR 2020, hundreds of citations), well past PyOD's inclusion bar in the [FAQ](https://pyod.readthedocs.io/en/latest/faq.html) (≥2 years since publication, 50+ citations, practical usefulness).
- There's recurring interest in semi-supervised detectors here (e.g. #527, #446). I didn't find a dedicated issue/PR for Deep SAD (only a passing mention in #334).

### Proposed implementation

- `pyod/models/deep_sad.py` — a `BaseDetector` subclass mirroring `deep_svdd.py`'s structure (same `fit` / `decision_function` contract, `decision_scores_` via `_process_decision_scores`), with `fit(X, y=None)` accepting optional labels so unlabeled samples fall back to the DeepSVDD-style objective and labeled ones use the Deep SAD loss.
- `examples/deep_sad_example.py`, `pyod/test/test_deep_sad.py`, and a README model-table row — matching the 3-file convention in the PR template.
- Reference implementations to adapt (both license-compatible with PyOD's BSD-2): the authors' [`lukasruff/Deep-SAD-PyTorch`](https://github.com/lukasruff/Deep-SAD-PyTorch) (MIT) and DeepOD's `dsad.py` (BSD-3).

Happy to author and maintain it. Does this sound good — and any preference on the interface (in particular how labeled vs. unlabeled samples should be passed to `fit`) before I start?

### PR DESCRIPTION
**Pull request by JayeshSuryavanshi**
## Add Deep SAD (Deep Semi-Supervised Anomaly Detection)

Closes #709.

This adds `DeepSAD`, the semi-supervised extension of the existing `DeepSVDD`, from Ruff et al., *Deep Semi-Supervised Anomaly Detection*, ICLR 2020.

### Motivation

PyOD ships `DeepSVDD` but not its widely-cited semi-supervised sibling. In fraud/risk and many real-world anomaly settings a small number of confirmed anomalies is available, and Deep SAD is the canonical way to use those labels: it keeps the Deep SVDD hypersphere objective for unlabeled data while pulling labeled anomalies *away* from the center, which measurably improves detection over the purely-unsupervised model.

### What it does

A network maps inputs into a representation space where inliers are pulled toward a fixed hypersphere center `c` and labeled anomalies are pushed away from it, via the Deep SAD objective:

- unlabeled samples use the Deep SVDD term, minimizing the squared distance to `c`;
- labeled anomalies use the inverse-distance term `(dist)^(-1)`, so their loss grows as they approach the center.

The distance to `c` is the anomaly score at inference. The semi-supervised interface is a single optional `y` in `fit(X, y=None)` where `y == 1` marks a known anomaly and everything else is unlabeled; `y=None` recovers unsupervised Deep SVDD, so it drops in wherever `DeepSVDD` is used.

### Implementation notes

- Follows PyOD conventions and mirrors `pyod/models/deep_svdd.py` (same `BaseDetector` contract, standardization, center initialization from a first forward pass, `decision_function` returning the distance-to-center score).
- The objective and center initialization are adapted from the authors' reference implementation (`lukasruff/Deep-SAD-PyTorch`, MIT) and DeepOD's `dsad.py` (BSD-3); neither is copied verbatim and both are attributed in the module header.

### Included

- `pyod/models/deep_sad.py` — the detector
- `pyod/test/test_deep_sad.py` — 15 tests mirroring the DeepSVDD test suite (parameters, scores, labels, proba, confidence, rejection, `fit_predict`, `clone`), all passing locally
- `examples/deep_sad_example.py`
- `README.rst` model table entry + `docs/zreferences.bib` citation

### Scope

This is the core detector plus docs/tests. I deliberately did **not** add DeepSAD to the `ad_engine` knowledge base (`algorithms.json`, `benchmarks.json`, routing rules, per-model analysis) because those entries carry curated metadata and benchmark numbers I don't want to fabricate; the skill/KB consistency checks pass as-is. Happy to add a proper KB entry as a follow-up if you'd like, or if you can point me at the benchmark source.

### CODE REVIEW

**Comment by JayeshSuryavanshi** [comment]
Thanks for the review. I addressed the two Deep-SAD-specific points in feb561a:

- **Encoder width**: removed the `hidden_neurons > n_features` check. Deep SAD uses only the encoder (no autoencoder branch like Deep SVDD), so wider hidden layers are valid; the check wrongly rejected datasets with fewer features than the default hidden width. Added a test that fits on 8-feature data with the default `[64, 32]`.
- **Center validation**: a user-supplied `c` is now validated against the representation dimension (`hidden_neurons[-1]`) and for finiteness, so a scalar or mismatched vector raises a clear error instead of silently broadcasting. Added tests for both the rejection and the valid path.

I intentionally left the other two suggestions (the `l2_regularizer` default weight decay, and seeding torch in `__init__` vs `fit`): both are inherited directly from the sibling `DeepSVDD`, and I'd rather keep Deep SAD consistent with it than diverge unilaterally here. Happy to change either if you'd prefer.

File: pyod/models/deep_sad.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+        if self.preprocessing:
+            self.scaler_ = StandardScaler()
+            X_norm = self.scaler_.fit_transform(X)
+        else:
+            X_norm = np.copy(X)
+
+        # Validate and complete the number of hidden neurons
+        if np.min(self.hidden_neurons) > self.n_features_:
+            raise ValueError("The number of neurons should not exceed "
+                             "the number of features")
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in feb561a: the width check now only rejects layers narrower than the previous one.

File: pyod/models/deep_sad.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+        X_norm = torch.tensor(X_norm, dtype=torch.float32)
+        semi_targets = torch.tensor(semi_targets, dtype=torch.float32)
+
+        # Initialize the hypersphere center from a first forward pass
+        if self.c is None:
+            self.c_ = self.model_._init_c(X_norm)
+        elif not torch.is_tensor(self.c):
+            self.c_ = torch.tensor(self.c, dtype=torch.float32)
+        else:
+            self.c_ = self.c
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in feb561a (width, finite) and 148f05f (all-zero centers rejected, c_ stored as a copy).

File: pyod/models/deep_sad.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+        rep_dim = self.hidden_neurons[-1]
+        if self.c is None:
+            self.c_ = self.model_._init_c(X_norm)
+        else:
+            if torch.is_tensor(self.c):
+                c_ = self.c.to(torch.float32)
+            else:
+                c_ = torch.tensor(self.c, dtype=torch.float32)
+            c_ = c_.flatten()
+            if c_.shape[0] != rep_dim:
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in 148f05f: a scalar c is expanded to hidden_neurons[-1].

File: pyod/models/deep_sad.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        self.validation_size = validation_size
+        self.preprocessing = preprocessing
+        self.eps = eps
+        self.verbose = verbose
+        self.random_state = random_state
+        self.model_ = None
+        self.best_model_dict = None
+
+        if self.random_state is not None:
+            torch.manual_seed(self.random_state)
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in 148f05f: a RandomState instance is turned into an int seed via check_random_state; int seeds are unchanged.

File: pyod/models/deep_sad.py [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+                epoch_loss += loss.item()
+            if epoch_loss < best_loss:
+                best_loss = epoch_loss
+                best_model_dict = self.model_.state_dict()
+            if self.verbose:
+                print(f"Epoch {epoch + 1}/{self.epochs}, "
+                      f"Loss: {epoch_loss}")
+        self.best_model_dict = best_model_dict
+
+        self.decision_scores_ = self.decision_function(X)
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py:490]: Addressed in 148f05f: the best state_dict is deep-copied and loaded back before decision_function.

File: pyod/models/deep_sad.py [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+    'sgd': optim.SGD,
+    'adam': optim.Adam,
+    'rmsprop': optim.RMSprop,
+    'adagrad': optim.Adagrad,
+    'adadelta': optim.Adadelta,
+    'adamw': optim.AdamW,
+    'nadam': optim.NAdam,
+    'sparseadam': optim.SparseAdam,
+    'asgd': optim.ASGD,
+    'lbfgs': optim.LBFGS
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in 148f05f: lbfgs and sparseadam removed from the mapping and the docstring.

File: pyod/test/test_deep_sad.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
+# -*- coding: utf-8 -*-
+
+
+import os
+import sys
+import unittest
+from unittest import mock
+
+import numpy as np
+import torch
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/test/test_deep_sad.py:10]: Addressed in 3b1c56c: test_deep_sad.py added to collect_ignore_glob in pyod/test/conftest.py.

File: pyod/models/deep_sad.py [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+        self.c = c
+        self.eta = eta
+        self.hidden_neurons = hidden_neurons or [64, 32]
+        self.hidden_activation = hidden_activation
+        self.optimizer = optimizer
+        self.epochs = epochs
+        self.batch_size = batch_size
+        self.dropout_rate = dropout_rate
+        self.l2_regularizer = l2_regularizer
+        self.validation_size = validation_size
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py:282]: Addressed in 3b1c56c: validation_size now holds out a split and the best epoch is chosen by validation loss; 0 falls back to training loss.

File: README.rst [hunk 9]
```diff
@@ -485,6 +485,11 @@ PyOD is organized into two functional groups: **(i) Detection Algorithms**, with
      - Deep One-Class Classification (`example <https://github.com/yzhao062/pyod/blob/development/examples/deepsvdd_example.py>`__)
      - 2018
      - [#Ruff2018Deep]_
+   * - Neural Networks
+     - DeepSAD
+     - Deep Semi-Supervised Anomaly Detection (`example <https://github.com/yzhao062/pyod/blob/development/examples/deep_sad_example.py>`__)
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply README.rst:490]: Addressed in 3b1c56c: DeepSAD row in docs/index.rst and an automodule block in docs/pyod.models.tabular.rst.

File: pyod/models/deep_sad.py [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+            for batch_x, batch_semi in dataloader:
+                optimizer.zero_grad()
+                outputs = self.model_(batch_x)
+                dist = torch.sum((outputs - self.c_) ** 2, dim=-1)
+                # Deep SAD loss: unlabeled/normal points (semi == 0)
+                # minimize the distance; labeled anomalies (semi == -1)
+                # minimize the inverse distance, pushing them away from c.
+                losses = torch.where(
+                    batch_semi == 0, dist,
+                    self.eta * ((dist + self.eps) ** batch_semi))
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in 3b1c56c: eta must be positive, checked in __init__.

File: pyod/test/test_deep_sad.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        self.semi_y = np.zeros(self.n_train, dtype=int)
+        anomaly_idx = np.where(self.y_train == 1)[0]
+        revealed = anomaly_idx[:len(anomaly_idx) // 2]
+        self.semi_y[revealed] = 1
+
+        self.clf = DeepSAD(n_features=self.n_features, epochs=10,
+                           hidden_neurons=[64, 32],
+                           contamination=self.contamination,
+                           random_state=2021)
+        self.clf.fit(self.X_train, self.semi_y)
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/test/test_deep_sad.py]: Addressed in 3b1c56c: fixtures moved to setUpClass; the file went from about 27s to 3s locally.

File: pyod/models/deep_sad.py [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+        return c
+
+    def _build_model(self):
+        layers = nn.Sequential()
+        layers.add_module('input_layer',
+                          nn.Linear(self.n_features, self.hidden_neurons[0],
+                                    bias=False))
+        layers.add_module('hidden_activation_e0',
+                          get_activation_by_name(self.hidden_activation))
+        for i in range(1, len(self.hidden_neurons) - 1):
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py:118]: Addressed in 3b1c56c: dropout now follows every hidden activation, the representation layer has none.

File: pyod/models/deep_sad.py [hunk 13]
```diff
@@ -0,0 +1,479 @@
...
+        semi_targets = torch.tensor(semi_targets, dtype=torch.float32)
+
+        # Hold out the validation split used for best-epoch selection; the
+        # center and the training batches only see the remaining samples.
+        n_val = int(self.n_samples_ * self.validation_size)
+        if n_val > 0:
+            perm = check_random_state(self.random_state).permutation(
+                self.n_samples_)
+            X_val, semi_val = X_norm[perm[:n_val]], semi_targets[perm[:n_val]]
+            X_fit, semi_fit = X_norm[perm[n_val:]], semi_targets[perm[n_val:]]
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in 61e36be: the validation split is stratified on the semi-supervised targets and never holds out all members of a class, so a lone labeled anomaly stays in training.

File: pyod/models/deep_sad.py [hunk 14]
```diff
@@ -0,0 +1,479 @@
...
+        self._classes = 2
+
+        self.n_samples_, self.n_features_ = X.shape[0], X.shape[1]
+
+        semi_targets = self._process_semi_targets(y, self.n_samples_)
+
+        # Standardize data for better performance
+        if self.preprocessing:
+            self.scaler_ = StandardScaler()
+            X_norm = self.scaler_.fit_transform(X)
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in 61e36be: the raw data is split first and the scaler is fitted on the training fold only.

File: pyod/models/deep_sad.py [hunk 15]
```diff
@@ -0,0 +1,479 @@
...
+
+        for epoch in range(self.epochs):
+            self.model_.train()
+            epoch_loss = 0
+            for batch_x, batch_semi in dataloader:
+                optimizer.zero_grad()
+                loss = self._deep_sad_loss(self.model_(batch_x), batch_semi)
+                loss.backward()
+                optimizer.step()
+                epoch_loss += loss.item()
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py]: Addressed in 61e36be: epoch loss is now the per-sample mean, each batch weighted by its size.

File: pyod/models/deep_sad.py [hunk 16]
```diff
@@ -0,0 +1,479 @@
...
+
+    def _deep_sad_loss(self, outputs, semi_targets):
+        """Deep SAD loss: unlabeled/normal points (target ``0``) minimize
+        the squared distance to the center; labeled anomalies (target
+        ``-1``) minimize the inverse distance, pushing them away from
+        ``c``.
+        """
+        dist = torch.sum((outputs - self.c_) ** 2, dim=-1)
+        losses = torch.where(semi_targets == 0, dist,
+                             self.eta * ((dist + self.eps) ** semi_targets))
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py:360]: Addressed in 61e36be: eps must be a positive finite float, checked in __init__.

File: pyod/models/deep_sad.py [hunk 17]
```diff
@@ -0,0 +1,479 @@
...
+            layers.add_module(f'hidden_layer_e{i}',
+                              nn.Linear(self.hidden_neurons[i - 1],
+                                        self.hidden_neurons[i], bias=False))
+            layers.add_module(f'hidden_activation_e{i}',
+                              get_activation_by_name(self.hidden_activation))
+            layers.add_module(f'hidden_dropout_e{i}',
+                              nn.Dropout(self.dropout_rate))
+        layers.add_module('net_output',
+                          nn.Linear(self.hidden_neurons[-2],
+                                    self.hidden_neurons[-1], bias=False))
```
[the remark this answers is not in the document]
Comment (JayeshSuryavanshi) [reply pyod/models/deep_sad.py:128]: Addressed in 61e36be: hidden_neurons=[k] builds a single bias-free linear layer into the representation space, with no activation or dropout.
